src/tcache.c

```diff
@@ -788,31 +796,33 @@ tcache_ncached_max_compute(szind_t szind, unsigned current_nhbins) {
 
 static void
 tcache_bin_info_compute(cache_bin_info_t *tcache_bin_info,
-    unsigned tcache_nhbins) {
-	for (szind_t i = 0; i < tcache_nhbins; i++) {
+    unsigned tcache_nbins) {
+	/*
+	 * Compute the values for each bin, but for bins with indices larger
+	 * than tcache_nbins, no items will be cached.
+	 */
+	for (szind_t i = 0; i < TCACHE_NBINS_MAX; i++) {
 		unsigned ncached_max = tcache_ncached_max_compute(i,
-		    tcache_nhbins);
+		    tcache_nbins);
+		assert(ncached_max <= CACHE_BIN_NCACHED_MAX);
 		cache_bin_info_init(&tcache_bin_info[i], ncached_max);
 	}
-	for (szind_t i = tcache_nhbins; i < SC_NBINS; i++) {
+	for (szind_t i = tcache_nbins; i < SC_NBINS; i++) {
 		/* Disabled small bins. */
 		cache_bin_info_init(&tcache_bin_info[i], 0);
```

Comment:
Question: it looks, now with this PR, the bins will look like this:
1) disabled small bins: ncached_max == 0, fake stack setup
2) disabled large bins: ncached_max non zero, NULL stack setup

They are a bit confusing and inconsistent. Maybe step back and think about how we should simply or unify things? (related to the comment about, if we should also set up large disabled bins with fake stacks)

Reply:
The design actually means to be:
1. disabled small bins: ncached_max non zero, fake stack setup
2. disabled large bins: ncached_max non zero, NULL stack setup.

This got a bit confuse mainly because: in the `tcache_small_bin_disabled`, we should check whether the `stack_head` equals to `cache_bin_preceding_junk` instead of simply checking `cache_bin_still_zero_initialized` through `cache_bin_info_ncached_max`. This is also why the previous `cache_bin_info_init` is not wrapped by an `else`. WDYT about this design? Or it's not that safe to use `cache_bin_preceding_junk` this way?

Reply:
Or maybe using `low_bits_empty == low_bits_full` as a unified way to determine whether the bin is disabled or not?

Reply:
> disabled small bins: ncached_max non zero,

This means we probably should remove this loop, which is setting ncached_max to 0 for disabled small bins?

> Or maybe using low_bits_empty == low_bits_full as a unified way to determine whether the bin is disabled or not?

This sounds good to me.

Reply:
It seems when the bin is stashed to full, it is possible to have `low_bits_empty == low_bits_full` while the bin is not init to a fake stack. Thus, I added `(uintptr_t)(*bin->stack_head) == cache_bin_preceding_junk` to make sure it's a fake stack. Yet I am not sure using the special value here is a good idea. Can discuss more offline regarding the design.

Reply:
That's a tricky case (glad the unit tests caught it). Looked through cache_bin_s again, and we rely on the ncached_max to determine the "true" full position, after flushing stashed items:
https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/cache_bin.h#L668

That means setting up the large disabled bins with fake stacks is confusing, because it would look similar to an enabled bin, full of stashed items, other than the magic value as you pointed out.

However *stack_head == cache_bin_preceding_junk is also not correct. bin[0] when empty will have head pointing there:
https://github.com/jemalloc/jemalloc/blob/dev/src/tcache.c#L691

So that leaves us with: 
```
if (ncached_max == 0) {
  return bin_disabled; /* could either be small or large */
}

if (stack_head == NULL) {
  assert(bin_is_large);
  return bin_disabled;
}

return bin_enabled;
```

Does the above sound right? Let's think a bit more about how to arrange these.

Reply:
> disabled small bins: ncached_max non zero, fake stack setup

How would this work? Does it also depend on if bin_index > tcache_nbins?

If bin_index < tcache_bins, ncached_max has to be zero to mean disabled, is that right? But we may allow ncached_max to be positive, if bin_index > tcache_bins?

Reply:
Discussed with @interwq offline. Now the design is: use a global pointer `disabled_bin` pointing to an address out of the stack, i.e., `&cache_bin_preceding_junk`. When initializing a disabled small bin, we initialize its `stack_head`, together with `empty_position` and `full_position` to `disabled_bin`. In this way, the fake stack fails in the fast path without an extra branch and we can determine whether a small bin is disabled swiftly while allowing users to set `ncached_max` to non-zero even the bin is still disabled.